narwhal/primary/src/synchronizer.rs

```diff
@@ -826,3 +859,163 @@ impl Synchronizer {
+                .entry((missing_round, d))
+                .or_default()
+                .insert(digest));
+        }
+        new_accept_notification(rx)
+    }
+
+    /// Examines children of a certificate that has been accepted, and returns the children that
+    /// can be accepted as well.
+    fn accept_children(
+        &mut self,
+        round: Round,
+        digest: CertificateDigest,
+    ) -> Vec<SuspendedCertificate> {
+        // This validation is only triggered for fetched and own certificates.
+        // Certificates from other sources will find the suspended certificate and wait on its
+        // accept notification, so no parent check or validation is done.
+        if let Some(suspended_cert) = self.suspended.remove(&digest) {
+            panic!(
+                "Suspended certificate {digest:?} has no missing parent ({:?} exist in store)",
+                suspended_cert.missing_parents
+            )
+        }
+        let mut to_traverse = VecDeque::new();
+        let mut to_accept = Vec::new();
+        to_traverse.push_back((round, digest));
+        while let Some((round, digest)) = to_traverse.pop_front() {
+            let Some(child_digests) = self.missing.remove(&(round, digest)) else {
+                continue;
+            };
+            for child in &child_digests {
+                let Some(suspended_cert) = self.suspended.get_mut(child) else {
+                    continue;
+                };
+                suspended_cert.missing_parents.remove(&digest);
+                if suspended_cert.missing_parents.is_empty() {
+                    let suspended_cert = self.suspended.remove(child).unwrap();
+                    to_traverse.push_back((
+                        suspended_cert.certificate.round(),
+                        suspended_cert.certificate.digest(),
+                    ));
+                    to_accept.push(suspended_cert);
+                }
+            }
+        }
+        to_accept
+    }
+
+    /// Runs GC on the suspended certificates, returns a list that can be accepted at gc round + 1.
+    /// It is caller's responsibility to check if some children of the returned certificates can
+    /// also be accepted.
+    fn run_gc(&mut self, gc_round: Round) -> Vec<SuspendedCertificate> {
+        // Remove suspended certificates below gc round, and collect digests for certificates just
+        // above the gc round.
+        let mut gc_certificates = Vec::new();
+        let mut certificates_above_gc_round = HashSet::new();
+        while let Some(((round, digest), children)) = self.missing.iter().next() {
+            if *round > gc_round {
+                break;
+            }
+            if *round == gc_round {
+                certificates_above_gc_round.extend(children.iter().cloned());
+            }
+            // It is ok to notify waiters here (via Drop). The certificate will never and does
+            // not need to get into certificate store.
```

Comment:
Why don't we want to store the certificates bellow GC round? I believe there is still value to do so in order to help later slow nodes catch up.

Reply:
If a certificate is not included in commits that raises the GC round above its own round, it will never get committed, so it is not useful to any node either.

Reply:
Sure yes, but it might still make them hit the "missing certificate" while they catch up. But I see the argument in any case and you are right that practically they are not needed.

Reply:
It is a good point that fetched certificates may get suspended. Even if we don't drop the certificate here, they can be dropped earlier in the handler / Synchronizer. Or consensus commit is delayed and GC round has not been updated for a certificate that will be accepted just above GC round. Updated certificate fetcher to tolerate errors from `try_accept_fetched_certificate()`.